Description: Honour compiler/linker option flags
Forwarded: pending
Author: Orion Poplawski <orion@cora.nwra.com>
diff -up a/nxcomp/configure.in b/nxcomp/configure.in
--- a/nxcomp/configure.in	2012-11-10 06:40:55.000000000 -0700
+++ b/nxcomp/configure.in	2012-12-11 13:13:19.390229196 -0700
@@ -7,8 +7,8 @@ AC_PREREQ(2.13)
 
 dnl Set our default compilation flags.
 
-CXXFLAGS="-O3 -fno-rtti -fno-exceptions"
-CFLAGS="-O3"
+CXXFLAGS="$CXXFLAGS -O3 -fno-rtti -fno-exceptions"
+CFLAGS="$CFLAGS -O3"
 
 dnl Reset default linking directives.
 
diff -up a/nxproxy/configure.in b/nxproxy/configure.in
--- a/nxproxy/configure.in	2012-11-10 06:40:55.000000000 -0700
+++ b/nxproxy/configure.in	2012-12-11 13:16:19.955301045 -0700
@@ -7,8 +7,8 @@ AC_PREREQ(2.13)
 
 dnl Reset default compilation flags.
 
-CXXFLAGS="-O3"
-CPPFLAGS="-O3"
+CXXFLAGS="$CXXFLAGS -O3"
+CPPFLAGS="$CPPFLAGS -O3"
 
 dnl Prefer headers and libraries from nx-X11 if present.
 
