--- a/nxcomp/Makefile.in
+++ b/nxcomp/Makefile.in
@@ -313,4 +313,4 @@
 
 distclean:	clean
 		-rm -rf autom4te.cache config.status config.log \
-		config.cache depend.status Makefile tags
+		config.cache depend.status Makefile tags configure
--- a/nxproxy/Makefile.in
+++ b/nxproxy/Makefile.in
@@ -109,4 +109,4 @@
 		$(PROGRAM) $(PROGRAM).exe $(LIBFULL) $(LIBLOAD) $(LIBSHARED) $(LIBARCHIVE)
 
 distclean:	clean
-		-rm -f config.status config.log config.cache depend.status Makefile tags
+		-rm -rf autom4te.cache config.status config.log config.cache depend.status Makefile tags configure
