Description: Fix FTBFS on ppc64 architecture
Author: Orion Poplawski <orion@cora.nwra.com>
diff -up a/nx-X11/lib/GL/mesa/main/Imakefile.inc b/nx-X11/lib/GL/mesa/main/Imakefile.inc
--- a/nx-X11/lib/GL/mesa/main/Imakefile.inc	2014-01-04 13:39:35.000000000 -0700
+++ b/nx-X11/lib/GL/mesa/main/Imakefile.inc	2014-01-24 12:04:00.085272615 -0700
@@ -1,5 +1,8 @@
 XCOMM $XFree86: xc/lib/GL/mesa/src/Imakefile.inc,v 1.7tsi Exp $
 
+/* gcc on ppc64 defines this, causes macro recursion with pixel below */
+#undef pixel
+
 #ifndef MesaBuildDir
 #define MesaBuildDir $(GLXLIBSRC)/mesa/main/
 #endif
