Description: Prevent underlinking by linking to libNX_X{11,damage,fixes).
Author: Mike Gabriel <mike.gabriel@das-netzwerkteam.de>

--- a/nx-X11/config/cf/lnxLib.tmpl
+++ b/nx-X11/config/cf/lnxLib.tmpl
@@ -61,6 +61,9 @@ XMULIB = $(XMULIBONLY) $(XTOOLLIB) $(XLI
 #define SharedXxf86vmReqs	$(LDPRELIB) $(XLIB)
 #define SharedxkbfileReqs	$(LDPRELIB) $(XONLYLIB)
 #define SharedxkbuiReqs $(LDPRELIB) $(XKBFILELIB) $(XONLYLIB)
+#define SharedXfixesReqs	$(LDPRELIB) $(XLIB)
+#define SharedXdamageReqs	$(LDPRELIB) $(XLIB)
+#define SharedXcompositeReqs	$(LDPRELIB) $(XDAMAGELIB) $(XFIXESLIB) $(XLIB)
 
 #else	/* UseElfFormat */
 
