Description: Do not build bundled libraries
Author: Orion Poplawski <orion@nwra.com>
--- a/nx-X11/config/cf/host.def	2013-08-30 13:23:26.000000000 -0600
+++ b/nx-X11/config/cf/host.def	2013-08-30 20:44:19.647460161 -0600
@@ -962,7 +962,7 @@ XCOMM $XFree86: xc/config/cf/xf86site.de
 
 #define SharedLibXdmcp         YES
 
-#define BuildXaw7              YES
+#define BuildXaw7              NO
 
 #else /* #if !defined(NXZaurusXServer) && !defined(NXiPAQXServer) && !defined(DarwinArchitecture) */
 
@@ -1023,3 +1023,8 @@ XCOMM $XFree86: xc/config/cf/xf86site.de
 
 #define BuildRman NO
 #define BuildFontEncLib NO
+#define BuildXcursorLibrary NO
+#define BuildXftLibrary NO
+#define BuildXft1Library NO
+#define SharedLibFont NO
+#define BuildLoadableXlibI18n NO
diff -up nx-libs-3.5.0.21/nx-X11/lib/Imakefile.bundled nx-libs-3.5.0.21/nx-X11/lib/Imakefile
--- a/nx-X11/lib/Imakefile	2013-08-30 08:18:40.000000000 -0600
+++ b/nx-X11/lib/Imakefile	2013-08-30 20:42:40.344850909 -0600
@@ -213,14 +213,8 @@ XEXTLIBDIR = Xext
 
 #if BuildLibraries
 OLDXLIBDIR = oldX
-ICELIBDIR = ICE
-SMLIBDIR = SM
-XTLIBDIR = Xt
-XMULIBDIR = Xmu
-XMUULIBDIR = Xmuu
 XPMLIBDIR = Xpm
 XTSTLIBDIR = Xtst
-FSLIBDIR = FS
 #endif
 
