Description: Fix several typos in nx-X11
Author: Mike Gabriel <mike.gabriel@das-netzwerkteam.de>

--- a/nx-X11/extras/Mesa/src/mesa/main/debug.c
+++ b/nx-X11/extras/Mesa/src/mesa/main/debug.c
@@ -43,7 +43,7 @@
    "GL_QUAD_STRIP",
    "GL_POLYGON",
    "outside begin/end",
-   "inside unkown primitive",
+   "inside unknown primitive",
    "unknown state"
 };
 
--- a/nx-X11/extras/Mesa/src/mesa/shader/arbprogparse.c
+++ b/nx-X11/extras/Mesa/src/mesa/shader/arbprogparse.c
@@ -4106,7 +4106,7 @@
    if (*inst++ != REVISION) {
       _mesa_set_program_error (ctx, 0, "Grammar version mismatch");
       _mesa_error(ctx, GL_INVALID_OPERATION,
-                  "glProgramStringARB(Grammar verison mismatch)");
+                  "glProgramStringARB(Grammar version mismatch)");
       err = GL_TRUE;
    }
    else {
--- a/nx-X11/ChangeLog.X.org
+++ b/nx-X11/ChangeLog.X.org
@@ -12733,7 +12733,7 @@
 	* programs/xkill/xkill.c: (get_window_id):
 	Changed cursor for the 'kill' action from XC_draped_box to
 	XC_Pirate. If you don't like it we can change it back
-	(original author unkown).
+	(original author unknown).
 
 	* programs/xman/vendor.h:
 	Added 'pic' to the man page rendering command pipeline
