--- a/nx-X11/extras/Mesa/src/mesa/drivers/dri/common/xmlconfig.c
+++ b/nx-X11/extras/Mesa/src/mesa/drivers/dri/common/xmlconfig.c
@@ -729,7 +729,7 @@
     for (i = 0; attr[i]; i += 2) {
 	if (!strcmp (attr[i], "driver")) driver = attr[i+1];
 	else if (!strcmp (attr[i], "screen")) screen = attr[i+1];
-	else XML_WARNING("unkown device attribute: %s.", attr[i]);
+	else XML_WARNING("unknown device attribute: %s.", attr[i]);
     }
     if (driver && strcmp (driver, data->driverName))
 	data->ignoringDevice = data->inDevice;
@@ -749,7 +749,7 @@
     for (i = 0; attr[i]; i += 2) {
 	if (!strcmp (attr[i], "name")) name = attr[i+1];
 	else if (!strcmp (attr[i], "executable")) exec = attr[i+1];
-	else XML_WARNING("unkown application attribute: %s.", attr[i]);
+	else XML_WARNING("unknown application attribute: %s.", attr[i]);
     }
     if (exec && strcmp (exec, data->execName))
 	data->ignoringApp = data->inApp;
@@ -762,7 +762,7 @@
     for (i = 0; attr[i]; i += 2) {
 	if (!strcmp (attr[i], "name")) name = attr[i+1];
 	else if (!strcmp (attr[i], "value")) value = attr[i+1];
-	else XML_WARNING("unkown option attribute: %s.", attr[i]);
+	else XML_WARNING("unknown option attribute: %s.", attr[i]);
     }
     if (!name) XML_WARNING1 ("name attribute missing in option.");
     if (!value) XML_WARNING1 ("value attribute missing in option.");
--- a/nx-X11/extras/Mesa/src/mesa/main/debug.c
+++ b/nx-X11/extras/Mesa/src/mesa/main/debug.c
@@ -43,7 +43,7 @@
    "GL_QUAD_STRIP",
    "GL_POLYGON",
    "outside begin/end",
-   "inside unkown primitive",
+   "inside unknown primitive",
    "unknown state"
 };
 
--- a/nx-X11/extras/Mesa/src/mesa/shader/arbprogparse.c
+++ b/nx-X11/extras/Mesa/src/mesa/shader/arbprogparse.c
@@ -4106,7 +4106,7 @@
    if (*inst++ != REVISION) {
       _mesa_set_program_error (ctx, 0, "Grammar version mismatch");
       _mesa_error(ctx, GL_INVALID_OPERATION,
-                  "glProgramStringARB(Grammar verison mismatch)");
+                  "glProgramStringARB(Grammar version mismatch)");
       err = GL_TRUE;
    }
    else {
--- a/nx-X11/extras/Mesa/src/mesa/shader/slang/Include/InfoSink.h
+++ b/nx-X11/extras/Mesa/src/mesa/shader/slang/Include/InfoSink.h
@@ -88,7 +88,7 @@
         case EPrefixError:         append("ERROR: ");          break;
         case EPrefixInternalError: append("INTERNAL ERROR: "); break;
         case EPrefixUnimplemented: append("UNIMPLEMENTED: ");  break;
-        default:                   append("UNKOWN ERROR: ");   break;
+        default:                   append("UNKNOWN ERROR: ");   break;
         }
     }
     void location(TSourceLoc loc) {
--- a/nx-X11/ChangeLog.X.org
+++ b/nx-X11/ChangeLog.X.org
@@ -12733,7 +12733,7 @@
 	* programs/xkill/xkill.c: (get_window_id):
 	Changed cursor for the 'kill' action from XC_draped_box to
 	XC_Pirate. If you don't like it we can change it back
-	(original author unkown).
+	(original author unknown).
 
 	* programs/xman/vendor.h:
 	Added 'pic' to the man page rendering command pipeline
