Index: Mesa/src/glx/x11/glcontextmodes.c
===================================================================
--- Mesa.orig/src/glx/x11/glcontextmodes.c
+++ Mesa/src/glx/x11/glcontextmodes.c
@@ -53,14 +53,14 @@
 # ifdef XFree86Server
 # include <os.h>
 # include <string.h>
-#  define _mesa_malloc(b) xalloc(b)
-#  define _mesa_free(m)   xfree(m)
+#  define _mesa_malloc(b) malloc(b)
+#  define _mesa_free(m)   free(m)
 #  define _mesa_memset   memset
 # else
 #  include <X11/Xlibint.h>
 #  define _mesa_memset memset
-#  define _mesa_malloc(b) Xmalloc(b)
-#  define _mesa_free(m) Xfree(m)
+#  define _mesa_malloc(b) malloc(b)
+#  define _mesa_free(m) free(m)
 # endif /* XFree86Server */
 #endif /* !defined(IN_MINI_GLX) */
 
Index: Mesa/src/mesa/drivers/x11/xm_image.c
===================================================================
--- Mesa.orig/src/mesa/drivers/x11/xm_image.c
+++ Mesa/src/mesa/drivers/x11/xm_image.c
@@ -49,7 +49,7 @@ XMesaImage *XMesaCreateImage(int bitsPer
 {
     XMesaImage *image;
 
-    image = (XMesaImage *)xalloc(sizeof(XMesaImage));
+    image = (XMesaImage *)malloc(sizeof(XMesaImage));
 
     if (image) {
 	image->width = width;
@@ -67,7 +67,7 @@ void XMesaDestroyImage(XMesaImage *image
 {
     if (image->data)
 	free(image->data);
-    xfree(image);
+    free(image);
 }
 
 unsigned long XMesaGetPixel(XMesaImage *image, int x, int y)
Index: Mesa/include/GL/xmesa_xf86.h
===================================================================
--- Mesa.orig/include/GL/xmesa_xf86.h
+++ Mesa/include/GL/xmesa_xf86.h
@@ -133,7 +133,7 @@ static _X_INLINE XMesaImage *XMesaGetIma
 					unsigned int height,
 					unsigned long plane_mask, int format)
 {
-    XMesaImage *img = Xcalloc(sizeof(*img));
+    XMesaImage *img = calloc(1, sizeof(*img));
 
     img->width = p->drawable.width;
     img->height = p->drawable.height;
