Index: Mesa/src/glx/x11/glcontextmodes.c
===================================================================
--- Mesa.orig/src/glx/x11/glcontextmodes.c
+++ Mesa/src/glx/x11/glcontextmodes.c
@@ -37,7 +37,7 @@
 #if defined(HAVE_DIX_CONFIG_H)
 # include <dix-config.h>
 #endif
-#include <X11/X.h>
+#include <nx-X11/X.h>
 #include <GL/glx.h>
 #include "GL/glxint.h"
 #endif
Index: Mesa/src/mesa/drivers/x11/glxheader.h
===================================================================
--- Mesa.orig/src/mesa/drivers/x11/glxheader.h
+++ Mesa/src/mesa/drivers/x11/glxheader.h
@@ -40,9 +40,9 @@
 
 #else
 
-# include <X11/Xlib.h>
-# include <X11/Xlibint.h>
-# include <X11/Xutil.h>
+# include <nx-X11/Xlib.h>
+# include <nx-X11/Xlibint.h>
+# include <nx-X11/Xutil.h>
 # ifdef USE_XSHM  /* was SHM */
 #  include <sys/ipc.h>
 #  include <sys/shm.h>
Index: Mesa/src/mesa/drivers/x11/xm_image.c
===================================================================
--- Mesa.orig/src/mesa/drivers/x11/xm_image.c
+++ Mesa/src/mesa/drivers/x11/xm_image.c
@@ -32,7 +32,7 @@ SOFTWARE OR THE USE OR OTHER DEALINGS IN
  */
 
 #include <stdlib.h>
-#include <X11/Xmd.h>
+#include <nx-X11/Xmd.h>
 
 #include "glxheader.h"
 #include "xmesaP.h"
